x/sysproxy/sysproxy_windows.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build windows
+
+package sysproxy
+
+import (
+	"errors"
+	"fmt"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/registry"
+)
+
+var (
+	modwininet            = windows.NewLazySystemDLL("wininet.dll")
+	procInternetSetOption = modwininet.NewProc("InternetSetOptionW")
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/wininet/option-flags
+// INTERNET_OPTION_SETTINGS_CHANGED: 39
+// Notifies the system that the registry settings have been changed so that it verifies the settings on the next call to InternetConnect.
+// INTERNET_OPTION_REFRESH: 37
+// Causes the proxy data to be reread from the registry for a handle. No buffer is required.
+// This option can be used on the HINTERNET handle returned by InternetOpen.
+// This is used by InternetSetOption.
+const (
+	INTERNET_OPTION_SETTINGS_CHANGED = 39
+	INTERNET_OPTION_REFRESH          = 37
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/api/wininet/nf-wininet-internetsetoptionw
+// internetSetOption sets an Internet option.
+func internetSetOption(hInternet uintptr, dwOption int, lpBuffer uintptr, dwBufferLength uint32) bool {
+	ret, _, _ := procInternetSetOption.Call(
+		hInternet,
+		uintptr(dwOption),
+		lpBuffer,
+		uintptr(dwBufferLength),
+	)
+	return ret != 0
+}
+
+func resetWininetProxySettings() error {
+
+	result1 := internetSetOption(0, INTERNET_OPTION_SETTINGS_CHANGED, 0, 0)
+	result2 := internetSetOption(0, INTERNET_OPTION_REFRESH, 0, 0)
+
+	if result1 && result2 {
+		fmt.Println("Operation successful")
+		return nil
+	} else {
+		fmt.Println("Operation failed")
+		return errors.New("Wininet setting change operation failed")
+	}
+}
+
+func SetProxy(ip string, port string) error {
+	key, err := registry.OpenKey(registry.CURRENT_USER, `Software\Microsoft\Windows\CurrentVersion\Internet Settings`, registry.SET_VALUE)
+	if err != nil {
+		return err
+	}
+	defer key.Close()
+
+	values := map[string]interface{}{
+		"MigrateProxy":  1,
```

Comment:
@jyyi1 Documentation for `MigrateProxy` DWORD was sparse. From what I understood by reading up one some resources I found ([1](https://learn.microsoft.com/en-us/troubleshoot/developer/browsers/administration/automatically-detect-settings-enabled), [2](https://www.pcreview.co.uk/threads/migrateproxy-registry.1871188/)), It appears that this is related to how Internet Explore / Edge migrate their proxy settings. 

I am a bit conflicted now since the docs say that IE sets this value to 1 after migrating the settings. BTW, I followed the way [GreenTunnel](https://github.com/SadeghHayeri/GreenTunnel/blob/master/src/utils/system-proxy.js) was setting up Windows proxy params and tested the code on Windows successfully. But there could be corner and untested cases that I missed. 

If you recall in the office, we looked at Chrome proxy settings was in sync with system proxy settings and we could see the settings switching in the Browser on Windows. I did not however test this with IE. I am looking further into this...

Reply:
Thanks for the information. I think we can ignore IE for now because Microsoft has already [ended its support on June 15, 2022](https://learn.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge).

For the Edge, since [it's using Chromium](https://support.microsoft.com/en-us/topic/microsoft-edge-chromium-1ce9507c-f09d-4de6-a706-eb52f46be90c), I believe it should behave the same as Chrome. 

Let's remove this setting and make sure everything works fine.

Reply:
@jyyi1 It seems like `"ProxyHttp1.1":  0,` setting is also related to IE according to [this](https://admx.help/?Category=InternetExplorer&Policy=Microsoft.Policies.InternetExplorer::Advanced_ProxyHttp1_1). I am going to remove this one too and test the changes on Windows.